File: backend/search_agent/schemas.py
"""Typed input/output contracts for the Search Agent.

All models serialize to camelCase JSON (the contract used by the future
frontend) while Python code uses snake_case attributes.
"""

from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel


class CamelModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


# ---------------------------------------------------------------------------
# Input contract
# ---------------------------------------------------------------------------

GoalType = Literal[
    "find_clients",
    "find_suppliers",
    "find_consultants",
    "find_partners",
    "find_investors",
    "find_hires",
]

SourceKind = Literal["web", "maps", "directories", "linkedin_public", "user_urls"]

ProviderName = Literal["mock", "exa", "tavily", "brave", "serper", "custom"]


class BusinessProfile(CamelModel):
    business_id: str
    business_name: str
    business_type: Optional[str] = None
    description: Optional[str] = None
    what_we_sell: str
    value_proposition: Optional[str] = None
    target_geographies: list[str] = Field(default_factory=list)
    ideal_customers: list[str] = Field(default_factory=list)
    bad_fit_customers: list[str] = Field(default_factory=list)
    preferred_tone: Optional[str] = None
    languages: list[str] = Field(default_factory=list)


class Mission(CamelModel):
    mission_id: str
    title: Optional[str] = None
    goal_type: GoalType
    description: str
    target_location: Optional[str] = None
    target_industry: Optional[str] = None
    target_business_size: Optional[str] = None
    desired_lead_count: Optional[int] = None
    urgency: Optional[Literal["low", "medium", "high"]] = None
    language: Optional[str] = None


class SearchOptions(CamelModel):
    max_queries: int = 6
    max_results_per_query: int = 8
    max_pages_to_fetch: int = 12
    include_sources: list[SourceKind] = Field(
        default_factory=lambda: ["web", "directories"]
    )
    user_provided_urls: list[str] = Field(default_factory=list)
    freshness_days: Optional[int] = None
    allow_llm: bool = False
    dry_run: bool = False


class ProviderOptions(CamelModel):
    provider: ProviderName = "mock"


class SearchAgentInput(CamelModel):
    request_id: str
    user_id: Optional[str] = None
    business_profile: BusinessProfile
    mission: Mission
    search_options: SearchOptions = Field(default_factory=SearchOptions)
    provider_options: ProviderOptions = Field(default_factory=ProviderOptions)


# ---------------------------------------------------------------------------
# Output contract
# ---------------------------------------------------------------------------

EvidenceType = Literal[
    "industry_match",
    "location_match",
    "service_match",
    "contact_found",
    "activity_signal",
    "bad_fit_signal",
    "other",
]

Category = Literal[
    "high_fit",
    "promising_but_incomplete",
    "needs_verification",
    "rejected_or_low_fit",
]

NextAction = Literal[
    "open_details",
    "investigate_more",
    "draft_outreach",
    "find_contact",
    "reject",
    "save_for_later",
]


class SearchPlan(CamelModel):
    interpreted_goal: str
    target_personas: list[str] = Field(default_factory=list)
    target_segments: list[str] = Field(default_factory=list)
    good_fit_signals: list[str] = Field(default_factory=list)
    bad_fit_signals: list[str] = Field(default_factory=list)
    suggested_sources: list[str] = Field(default_factory=list)
    generated_queries: list[str] = Field(default_factory=list)
    assumptions: list[str] = Field(default_factory=list)


class EmailContact(CamelModel):
    value: str
    type: Literal["generic", "personal", "unknown"] = "unknown"
    confidence: float = 0.5
    source_url: Optional[str] = None


class PhoneContact(CamelModel):
    value: str
    confidence: float = 0.5
    source_url: Optional[str] = None


class SocialLink(CamelModel):
    platform: Literal["linkedin", "facebook", "instagram", "x", "other"]
    url: str


class ContactInfo(CamelModel):
    emails: list[EmailContact] = Field(default_factory=list)
    phones: list[PhoneContact] = Field(default_factory=list)
    contact_page_url: Optional[str] = None
    social_links: list[SocialLink] = Field(default_factory=list)


class Evidence(CamelModel):
    source_url: str
    title: Optional[str] = None
    snippet: str
    evidence_type: EvidenceType
    confidence: float = 0.5


class Scores(CamelModel):
    fit_score: float = 0
    contactability_score: float = 0
    evidence_quality_score: float = 0
    freshness_score: Optional[float] = None
    overall_score: float = 0


class Classification(CamelModel):
    category: Category
    confidence: Literal["low", "medium", "high"] = "medium"
    reasons: list[str] = Field(default_factory=list)
    missing_info: list[str] = Field(default_factory=list)
    recommended_next_action: NextAction = "open_details"


class CandidateLead(CamelModel):
    id: str
    mission_id: str
    name: str
    type: Literal["company", "person", "unknown"] = "company"
    website_url: Optional[str] = None
    domain: Optional[str] = None
    source_urls: list[str] = Field(default_factory=list)
    source_provider: Optional[str] = None
    source_query: Optional[str] = None

    short_description: Optional[str] = None
    location: Optional[str] = None
    industry: Optional[str] = None
    business_size_estimate: Optional[str] = None

    contact: ContactInfo = Field(default_factory=ContactInfo)
    evidence: list[Evidence] = Field(default_factory=list)
    scores: Scores = Field(default_factory=Scores)
    classification: Classification = Field(
        default_factory=lambda: Classification(category="needs_verification")
    )

    created_at: str
    updated_at: str


class Groups(CamelModel):
    high_fit: list[str] = Field(default_factory=list)
    promising_but_incomplete: list[str] = Field(default_factory=list)
    needs_verification: list[str] = Field(default_factory=list)
    rejected_or_low_fit: list[str] = Field(default_factory=list)


class Summary(CamelModel):
    queries_run: int = 0
    raw_results_found: int = 0
    pages_fetched: int = 0
    candidates_created: int = 0
    duplicates_removed: int = 0
    high_fit_count: int = 0
    warning_count: int = 0


class SearchWarning(CamelModel):
    code: str
    message: str
    severity: Literal["info", "warning", "error"] = "warning"
    related_lead_id: Optional[str] = None


class SearchTraceEvent(CamelModel):
    timestamp: str
    step: str
    message: str
    metadata: Optional[dict] = None


class SearchAgentOutput(CamelModel):
    request_id: str
    mission_id: str
    status: Literal["success", "partial_success", "failed"]
    search_plan: SearchPlan
    candidates: list[CandidateLead] = Field(default_factory=list)
    groups: Groups = Field(default_factory=Groups)
    summary: Summary = Field(default_factory=Summary)
    warnings: list[SearchWarning] = Field(default_factory=list)
    trace: list[SearchTraceEvent] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# View model for the future Discover page
# ---------------------------------------------------------------------------


class DiscoverLeadCardVM(CamelModel):
    id: str
    title: str
    subtitle: Optional[str] = None
    location: Optional[str] = None
    domain: Optional[str] = None
    contact_badges: list[str] = Field(default_factory=list)
    fit_score: float = 0
    category: str = "needs_verification"
    reasons: list[str] = Field(default_factory=list)
    missing_info: list[str] = Field(default_factory=list)
    recommended_next_action: str = "open_details"
    primary_action_label: str = "Open details"

File: src/search-agent/schemas.ts
import { z } from "zod";

export const BusinessProfileSchema = z.object({
  businessId: z.string(),
  businessName: z.string(),
  businessType: z.string().optional(),
  description: z.string().optional(),
  whatWeSell: z.string(),
  valueProposition: z.string().optional(),
  targetGeographies: z.array(z.string()).optional(),
  idealCustomers: z.array(z.string()).optional(),
  badFitCustomers: z.array(z.string()).optional(),
  preferredTone: z.string().optional(),
  languages: z.array(z.string()).optional(),
});

export const MissionSchema = z.object({
  missionId: z.string(),
  title: z.string().optional(),
  goalType: z.enum([
    "find_clients",
    "find_suppliers",
    "find_consultants",
    "find_partners",
    "find_investors",
    "find_hires",
  ]),
  description: z.string(),
  targetLocation: z.string().optional(),
  targetIndustry: z.string().optional(),
  targetBusinessSize: z.string().optional(),
  desiredLeadCount: z.number().optional(),
  urgency: z.enum(["low", "medium", "high"]).optional(),
  language: z.string().optional(),
});

export const SearchOptionsSchema = z.object({
  maxQueries: z.number().optional(),
  maxResultsPerQuery: z.number().optional(),
  maxPagesToFetch: z.number().optional(),
  includeSources: z
    .array(
      z.enum(["web", "maps", "directories", "linkedin_public", "user_urls"]),
    )
    .optional(),
  userProvidedUrls: z.array(z.string()).optional(),
  freshnessDays: z.number().optional(),
  allowLLM: z.boolean().optional(),
  dryRun: z.boolean().optional(),
});

export const ProviderOptionsSchema = z.object({
  provider: z
    .enum(["mock", "exa", "tavily", "brave", "serper", "custom"])
    .optional(),
});

export const SearchAgentInputSchema = z.object({
  requestId: z.string(),
  userId: z.string().optional(),
  businessProfile: BusinessProfileSchema,
  mission: MissionSchema,
  searchOptions: SearchOptionsSchema.optional(),
  providerOptions: ProviderOptionsSchema.optional(),
});

export const EmailContactSchema = z.object({
  value: z.string(),
  type: z.enum(["generic", "personal", "unknown"]),
  confidence: z.number(),
  sourceUrl: z.string().optional(),
});

export const PhoneContactSchema = z.object({
  value: z.string(),
  confidence: z.number(),
  sourceUrl: z.string().optional(),
});

export const SocialLinkSchema = z.object({
  platform: z.enum(["linkedin", "facebook", "instagram", "x", "other"]),
  url: z.string(),
});

export const ContactSchema = z.object({
  emails: z.array(EmailContactSchema),
  phones: z.array(PhoneContactSchema),
  contactPageUrl: z.string().optional(),
  socialLinks: z.array(SocialLinkSchema).optional(),
});

export const EvidenceSchema = z.object({
  sourceUrl: z.string(),
  title: z.string().optional(),
  snippet: z.string(),
  evidenceType: z.enum([
    "industry_match",
    "location_match",
    "service_match",
    "contact_found",
    "activity_signal",
    "bad_fit_signal",
    "other",
  ]),
  confidence: z.number(),
});

export const ScoresSchema = z.object({
  fitScore: z.number(),
  contactabilityScore: z.number(),
  evidenceQualityScore: z.number(),
  freshnessScore: z.number().optional(),
  overallScore: z.number(),
});

export const ClassificationSchema = z.object({
  category: z.enum([
    "high_fit",
    "promising_but_incomplete",
    "needs_verification",
    "rejected_or_low_fit",
  ]),
  confidence: z.enum(["low", "medium", "high"]),
  reasons: z.array(z.string()),
  missingInfo: z.array(z.string()),
  recommendedNextAction: z.enum([
    "open_details",
    "investigate_more",
    "draft_outreach",
    "find_contact",
    "reject",
    "save_for_later",
  ]),
});

export const CandidateLeadSchema = z.object({
  id: z.string(),
  missionId: z.string(),
  name: z.string(),
  type: z.enum(["company", "person", "unknown"]),
  websiteUrl: z.string().optional(),
  domain: z.string().optional(),
  sourceUrls: z.array(z.string()),
  sourceProvider: z.string().optional(),
  sourceQuery: z.string().optional(),
  shortDescription: z.string().optional(),
  location: z.string().optional(),
  industry: z.string().optional(),
  businessSizeEstimate: z.string().optional(),
  contact: ContactSchema,
  evidence: z.array(EvidenceSchema),
  scores: ScoresSchema,
  classification: ClassificationSchema,
  createdAt: z.string(),
  updatedAt: z.string(),
});

export const SearchPlanSchema = z.object({
  interpretedGoal: z.string(),
  targetPersonas: z.array(z.string()),
  targetSegments: z.array(z.string()),
  goodFitSignals: z.array(z.string()),
  badFitSignals: z.array(z.string()),
  suggestedSources: z.array(z.string()),
  generatedQueries: z.array(z.string()),
  assumptions: z.array(z.string()),
});

export const SearchWarningSchema = z.object({
  code: z.string(),
  message: z.string(),
  severity: z.enum(["info", "warning", "error"]),
  relatedLeadId: z.string().optional(),
});

export const SearchTraceEventSchema = z.object({
  timestamp: z.string(),
  step: z.string(),
  message: z.string(),
  metadata: z.record(z.unknown()).optional(),
});

export const SearchAgentOutputSchema = z.object({
  requestId: z.string(),
  missionId: z.string(),
  status: z.enum(["success", "partial_success", "failed"]),
  searchPlan: SearchPlanSchema,
  candidates: z.array(CandidateLeadSchema),
  groups: z.object({
    highFit: z.array(z.string()),
    promisingButIncomplete: z.array(z.string()),
    needsVerification: z.array(z.string()),
    rejectedOrLowFit: z.array(z.string()),
  }),
  summary: z.object({
    queriesRun: z.number(),
    rawResultsFound: z.number(),
    pagesFetched: z.number(),
    candidatesCreated: z.number(),
    duplicatesRemoved: z.number(),
    highFitCount: z.number(),
    warningCount: z.number(),
  }),
  warnings: z.array(SearchWarningSchema),
  trace: z.array(SearchTraceEventSchema),
});

export const DiscoverLeadCardVMSchema = z.object({
  id: z.string(),
  title: z.string(),
  subtitle: z.string().optional(),
  location: z.string().optional(),
  domain: z.string().optional(),
  contactBadges: z.array(z.string()),
  fitScore: z.number(),
  category: z.string(),
  reasons: z.array(z.string()),
  missingInfo: z.array(z.string()),
  recommendedNextAction: z.string(),
  primaryActionLabel: z.string(),
});

export type SearchAgentInput = z.infer<typeof SearchAgentInputSchema>;
export type SearchAgentOutput = z.infer<typeof SearchAgentOutputSchema>;
export type CandidateLead = z.infer<typeof CandidateLeadSchema>;
export type SearchPlan = z.infer<typeof SearchPlanSchema>;
export type SearchWarning = z.infer<typeof SearchWarningSchema>;
export type SearchTraceEvent = z.infer<typeof SearchTraceEventSchema>;
export type DiscoverLeadCardVM = z.infer<typeof DiscoverLeadCardVMSchema>;
export type BusinessProfile = z.infer<typeof BusinessProfileSchema>;
export type Mission = z.infer<typeof MissionSchema>;
